RATapi/classlist.py

```diff
@@ -466,43 +466,34 @@ def __get_pydantic_core_schema__(cls, source: Any, handler):
 
         # if annotated with a class, get the item type of that class
         origin = get_origin(source)
-        if origin is None:
-            origin = source
-            item_tp = Any
-        else:
-            item_tp = get_args(source)[0]
+        item_tp = Any if origin is None else get_args(source)[0]
 
-        item_schema = handler.generate_schema(list[item_tp])
+        list_schema = handler.generate_schema(list[item_tp])
 
         def coerce(v: Any, handler: ValidatorFunctionWrapHandler) -> ClassList[T]:
-            """If a list is given, try to coerce it to a ClassList."""
-            if isinstance(v, list):
-                v = handler(v)
-                v = ClassList(v)
+            """If a sequence is given, try to coerce it to a ClassList."""
+            if isinstance(v, Sequence):
+                classlist = ClassList()
+                if len(v) > 0 and isinstance(v[0], dict):
+                    # we want to be OK if the type is a model and is passed as a dict;
+                    # pydantic will coerce it or fall over later
+                    classlist._class_handle = dict
+                elif item_tp is not Any:
+                    classlist._class_handle = item_tp
+                classlist.extend(v)
+                v = classlist
             return v
 
         def validate_items(v: ClassList[T], handler: ValidatorFunctionWrapHandler) -> ClassList[T]:
             v.data = handler(v.data)
             return v
 
-        validation_schema = core_schema.chain_schema(
+        schema = core_schema.chain_schema(
             [
-                core_schema.no_info_wrap_validator_function(coerce, item_schema),
+                core_schema.no_info_wrap_validator_function(coerce, list_schema),
```

Comment:
You use a wrap validator function here, but (now) don't call the handler within the `coerce` function. I'd recommend replacing this line and the next with `core_schema.no_info_before_validator_function(coerce, core_schema.is_instance_schema(cls))` - i.e., coerce the input if necessary and then check it's a ClassList in a single line.